FeedAPIChallenge/Feed API/RemoteFeedLoader.swift

```diff
@@ -13,12 +20,53 @@ public final class RemoteFeedLoader: FeedLoader {
 		case invalidData
 	}
 
+	public typealias Result = FeedLoader.Result
+
 	public init(url: URL, client: HTTPClient) {
 		self.url = url
 		self.client = client
 	}
 
 	public func load(completion: @escaping (FeedLoader.Result) -> Void) {
-		fatalError("Must be implemented")
+		client.get(from: url) { [weak self] result in
+			guard self != nil else { return }
+
+			switch result {
+			case let .success((data, response)): completion(RemoteFeedLoader.map(data, from: response))
+			case .failure:
+				completion(.failure(Error.connectivity))
+			}
+		}
+	}
+
+	private static func map(_ data: Data, from response: HTTPURLResponse) -> Result {
+		do {
+			let items = try FeedImagesMapper.map(data, from: response)
+			return .success(items.toModels())
+		} catch {
+			return .failure(error)
+		}
+	}
+
+	private class FeedImagesMapper {
+		private static var OK_200: Int { return 200 }
+
+		private struct Root: Decodable {
+			let items: [Image]
+		}
+
+		static func map(_ data: Data, from response: HTTPURLResponse) throws -> [Image] {
+			guard response.statusCode == OK_200, let root = try? JSONDecoder().decode(Root.self, from: data)
+			else {
+				throw RemoteFeedLoader.Error.invalidData
+			}
+			return root.items
+		}
```

Comment:
This method can return not `[Image]` but `[FeedImage]` and do all the mapping inline using `Array.map`.
Thus `Image` can be placed inside `FeedImagesMapper` the same way as `Root` is placed and `Array` extension on lines `68`-`72` can be removed. 
The code would feel a bit cleaner as all the remote models are placed in the same namespace.

```suggestion
		static func map(_ data: Data, from response: HTTPURLResponse) throws -> [FeedImage] {
			guard response.statusCode == OK_200, let root = try? JSONDecoder().decode(Root.self, from: data)
			else {
				throw RemoteFeedLoader.Error.invalidData
			}
			return root.items.map { FeedImage(id: $0.image_id, description: $0.image_desc, location: $0.image_loc, url: $0.image_url) }
		}
```

Reply:
Perfect. It's done. Thank you very much !